server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java

```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+        internalCluster().startNodes(3, settings);
+
+        final String stoppedNodeName = createDanglingIndices(INDEX_NAME, OTHER_INDEX_NAME);
+        final String danglingIndexUUID = findDanglingIndexForNode(stoppedNodeName, INDEX_NAME);
+
+        client().admin().cluster().deleteDanglingIndex(new DeleteDanglingIndexRequest(danglingIndexUUID, true)).actionGet();
+
+        assertThat(listDanglingIndices(), is(empty()));
+    }
+
+    /**
+     * Check that when a index is found to be dangling on more than one node, it can be deleted.
+     */
+    public void testDanglingIndexOverMultipleNodesCanBeDeleted() throws Exception {
+        final Settings settings = buildSettings(1, true, false);
+        internalCluster().startNodes(3, settings);
+
+        createIndices(INDEX_NAME, OTHER_INDEX_NAME);
+
+        // Restart 2 nodes, deleting the indices in their absence, so that there is a dangling index to recover
         internalCluster().restartRandomDataNode(new InternalTestCluster.RestartCallback() {
 
             @Override
             public Settings onNodeStopped(String nodeName) throws Exception {
-                internalCluster().validateClusterFormed();
-                assertAcked(client().admin().indices().prepareDelete(INDEX_NAME));
+                internalCluster().restartRandomDataNode(new InternalTestCluster.RestartCallback() {
+
+                    @Override
+                    public Settings onNodeStopped(String nodeName) throws Exception {
+                        internalCluster().validateClusterFormed();
+                        assertAcked(client().admin().indices().prepareDelete(INDEX_NAME));
+                        assertAcked(client().admin().indices().prepareDelete(OTHER_INDEX_NAME));
+                        return super.onNodeStopped(nodeName);
+                    }
+                });
+
                 return super.onNodeStopped(nodeName);
             }
         });
 
-        // Since index recovery is async, we can't prove index recovery will never occur, just that it doesn't occur within some reasonable
-        // amount of time
-        assertFalse(
-            "Did not expect dangling index " + INDEX_NAME + " to be recovered",
-            waitUntil(() -> indexExists(INDEX_NAME), 1, TimeUnit.SECONDS)
+        final AtomicReference<List<DanglingIndexInfo>> danglingIndices = new AtomicReference<>();
+
+        final List<DanglingIndexInfo> results = listDanglingIndices();
+
+        // Both the stopped nodes should have found a dangling index.
+        assertThat(results, hasSize(2));
+        danglingIndices.set(results);
+
+        // Try to delete the index - this request should succeed
+        client().admin()
+            .cluster()
+            .deleteDanglingIndex(new DeleteDanglingIndexRequest(danglingIndices.get().get(0).getIndexUUID(), true))
+            .actionGet();
+
+        assertBusy(() -> assertThat(listDanglingIndices(), empty()));
```

Comment:
Also here -- WDYT about making the same fix to this test as the REST test to verify that the index really did get deleted?